### PR TITLE
[quant][graphmode] Default to non-inplace in graph mode quantization API

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32379 [WIP][quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes
* #32380 [quant][graphmode][refactor] Better API for fold_convbn
* #32374 [refactor] Adding FoldConvBatchNorm2dHelper
* #32303 [quant][graphmode] Call _jit_pass_dedup_module_ueses in quantize_script
* #32302 [quant] Re-enable fold_convbn in quantize_script
* #32206 [quant] Re-enable test_nested that has different qconfig for shared ClassType
* #32205 [quant][graphmode] Support quantizing shared ClassType with different qconfigs
* **#32204 [quant][graphmode] Default to non-inplace in graph mode quantization API**
* #32203 Adding QConfigTypePtrMap
* #32202 [refactor] Move some of the helper functions for public use
* #31408 [quant][graphmode] `insert_quant_dequant` pass support shared class types

Summary:
att

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19508030](https://our.internmc.facebook.com/intern/diff/D19508030)

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 6c1c6bd1:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=6c1c6bd1558c873daade9745b8d378fe4d358ed3).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4248388) pytorch_xla_linux_xenial_py3_6_clang7_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86222654) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86222654))
<details>
<summary>
<code>Jan 15 02:39:47 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff</code>
</summary>

```
Jan 15 02:39:47 + XDIR=/var/lib/jenkins/workspace/xla/scripts/.. 
Jan 15 02:39:47 + PTDIR=/var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 15 02:39:47 + python /var/lib/jenkins/workspace/xla/scripts/cond_patch.py /var/lib/jenkins/workspace/xla/scripts/../torch_patches /var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 15 02:39:47 1 out of 1 hunk FAILED 
Jan 15 02:39:47 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 15 02:39:47 Traceback (most recent call last): 
Jan 15 02:39:47   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 67, in <module> 
Jan 15 02:39:47     patch_repo(args) 
Jan 15 02:39:47   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 49, in patch_repo 
Jan 15 02:39:47     raise RuntimeError('Failed to apply patch: {}'.format(ppath)) 
Jan 15 02:39:47 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 15 02:39:47 + cleanup 
Jan 15 02:39:47 + retcode=1 
Jan 15 02:39:47 + set +x 
Jan 15 02:39:47 =================== sccache compilation log =================== 
Jan 15 02:39:47 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 15 02:39:47 Compile requests                 1 
Jan 15 02:39:47 Compile requests executed        0 
Jan 15 02:39:47 Cache hits                       0 
Jan 15 02:39:47 Cache misses                     0 
Jan 15 02:39:47 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: torch/quantization/_quantize_script.py [hunk 1]
```diff
@@ -69,17 +70,17 @@ def prepare_script(model, qconfig_dict, inplace=False):
     _check_is_script_module(model)
     if not inplace:
         model = model.copy()
```
Comment (ZolotukhinM) [inline torch/quantization/_quantize_script.py]: That copy above seems redundant then.

File: torch/quantization/_quantize_script.py [hunk 2]
```diff
@@ -103,13 +104,7 @@ def quantize_script(model, qconfig_dict, run_fn, run_args, inplace=False):
     # revisit after constants is properly handled in
     # JIT
     # torch._C._jit_pass_fold_convbn(model._c)
-    prepare_script(model, scripted_qconfig_dict, True)
+    model = prepare_script(model, scripted_qconfig_dict, True)
```
Comment (ZolotukhinM) [inline torch/quantization/_quantize_script.py]: Should the last argument also be `False`?

File: torch/quantization/_quantize_script.py [hunk 3]
```diff
@@ -103,13 +104,7 @@ def quantize_script(model, qconfig_dict, run_fn, run_args, inplace=False):
...
     run_fn(model._c._get_method('forward'), *run_args)
-    # When we mutating graph we didn't create a new ClassType
-    # and the graph executor will run an out dated version
-    # of the graph if we do inplace graph mutation, therefore
-    # we copy the model here
-    # [TODO] This will be fixed later when we figure out
-    # how to properly mutate types
-    model = convert_script(model, False)
+    model = convert_script(model, True)
```
Comment (ZolotukhinM) [inline torch/quantization/_quantize_script.py]: Here as well.

**Review by ZolotukhinM — APPROVED** [review]
👍 
Some comments are inline.

File: torch/quantization/_quantize_script.py (hunk 1 above)
Comment (jerryzh168) [reply torch/quantization/_quantize_script.py]: yeah, I'll do another refactor to remove `inplace` argument

File: torch/quantization/_quantize_script.py (hunk 2 above)
Comment (jerryzh168) [reply torch/quantization/_quantize_script.py]: I'll do another refactor to remove this argument

**Pull request closed without merging**